shadowsocks/cipher_list.go

```diff
@@ -33,18 +38,21 @@ type CipherEntry struct {
 // CipherList is a thread-safe collection of CipherEntry elements that allows for
 // snapshotting and moving to front.
 type CipherList interface {
-	SnapshotForClientIP(clientIP net.IP) []*list.Element
+	// Returns a snapshot of the cipher list optimized for this client IP,
+	// and also the number of bytes needed for TCP trial decryption.
+	SnapshotForClientIP(clientIP net.IP) (int, []*list.Element)
```

Comment:
The CipherList API is getting a bit awkward. Why do we return a TCP-specific value in something used by UDP? I think that's because part of the cipher finding logic is here and part is in tcp.go and udp.go, and the boundary is a bit arbitrary. Maybe they belong together, something to think about.

Maybe we should rename it CipherFinder, and have three methods: Update, findCipherTCP and findCipherUDP. The other two would be hidden. Or even a `findCipher(buffer, testCipher, clientIP)`, that lets you pass `TestTCP(cipher, buffer)` or `TestUDP(cipher, buffer)`, which test if the buffer successfully decrypts for TCP or UDP.

But we don't need to change that for this PR.

Reply:
That does seem like an improvement.

BTW, I just realized that all UDP trial decryption (which I expect is the great majority) happens on a single thread.  We might be able to optimize CipherList by making it _not_ thread-safe (no locking), and instead just have one copy for TCP and one for UDP.  (However, we would also have to rethink access key updates.)  Alternatively, we could spread UDP handling across multiple threads.